Refactor beat support

KnechtionsCoding opened the pull request:
Per your request this supports the other beats and keeps the original elk-client.yml structure.

sadsfae commented:
@KnechtionsCoding some great work here.  I was on the fence about EL6 support also as it's coming up on 7years old and worried about supporting too much.

Just some small fixes in this PR:

1) ```- { role: instructions }``` needs to be the last role in the ```elk.yml``` playbook as it's job is solely print messages to people for next steps (initializing/picking your index, command to run elk-client playbook etc).

2) ```ansible-playbook -i hosts install/elk-client.yml --extra-vars 'elk_server=X.X.X.X``` really needs to be run as a second step here (you need index initialized first and we want people to choose what kind of and name of indices they want).

There is purposeful separation of ```install/elk.yml``` and ```install/elk-client.yml``` here for this reason as once the stack is deployed you print next steps, next step being run the elk-client playbook (where your *beat stuff can live).  I think all the beats logic can go into ```elk-client.yml``` or call roles outside of that but don't call them in ```elk.yml```.  The configuration options you have ```install/group_vars/all.yml``` are great though.

3) I'm getting some errors with the SSL filebeat cert pull from elk_server var (this is why elk_client and install/elk-client.yml has to be run as a second step since you're passing the -e var of your ELK server in order to know where to get the certificate.  Fixing 2 above should rectify this.

```
TASK [elk_client : Install ELK server SSL client certificate] ****************************************************************************************************************
fatal: [host-01]: FAILED! => {"failed": true, "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'elk_server' is undefined\n\nThe error appears to have been in '/home/wfoster/git_modules/ansible-elk/install/roles/elk_client/tasks/main.yml': line 29, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Install ELK server SSL client certificate\n  ^ here\n"}
```

Overall this looks great just a few minor things and I'd love to get this merged.  Thanks again, great work!

knechtionscoding commented:
@sadsfae I made those changes. I completely blanked on that. In the other version heartbeat was part of the elk_client setup steps. Not sure why I moved it over to elk other than conceptually it should be run at the same time because it isn't installed on lots of machines. Rather it's the one beat that is centralized. So the instructions for it needs to be clear that it should be installed on a centralized server (i use the ELK server itself).

And I'm more then happy to help support EL6 on this moving forward as I'll be using this playbook so much anyway.

sadsfae commented:
On 07/09/17 13:04 +0000, Sven Hans Knecht wrote:
>   [1]@sadsfae I made those changes. I completely blanked on that. In the
>   other version heartbeat was part of the elk_client setup steps. Not
>   sure why I moved it over to elk other than conceptually it should be
>   run at the same time because it isn't installed on lots of machines.
>   Rather it's the one beat that is centralized. So the instructions for
>   it needs to be clear that it should be installed on a centralized
>   server (i use the ELK server itself).
>
>   And I'm more then happy to help support EL6 on this moving forward as
>   I'll be using this playbook so much anyway.
>
>   —

Hey that's cool.  Ideally I'd love to support more than the latest
CentOS/RHEL but if you can continue to help that'd be wonderful.

I think we'll make the cutoff on only EL-based distributions however,
I know there is an issue for Ubuntu/Debian support as well but I don't
use it and certainly won't have the time to support it unfortunately.

sadsfae commented:
@KnechtionsCoding got farther this time, I think you're missing a defined variable.
```
TASK [metricbeat : Generate metricbeat configuration template] ***************************************************************************************************************
fatal: [host-02]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'logstash_beats_port' is undefined"}
```

I think ```logstash_beats_port``` needs to be added to group_vars/all.yml
```
32     hosts: ["{{ elk_server }}:{{ logstash_beats_port }}"]
```
You'll want to open up this port via the iptables/firewalld templates as well here on the ELK server if it's enabled via ```when: install_metricbeat```

You could put this in the elasticsearch role like this:

https://github.com/sadsfae/ansible-elk/blob/master/install/roles/elasticsearch/tasks/main.yml#L106:L142

Replace ```es_listen_external ``` with ``` install_metricbeat``` and include the other adaptive firewall (hack) code :)

sadsfae commented:
@KnechtionsCoding One other thing, if you're including CentOS/RHEL6 support you'll also want a conditional to get the right epel-release RPM/repo in the nginx role.

We can move this from a defined ```{{epel_repo}}``` to perhaps a jinja2 template that constructions the .repo based on whether it's EL6 or EL7 but there are lots of ways to do it.

You might be able to simply variabilize the repo url path also directly in the playbook.

Something like this might work:

```ansible_distribution_major_version|int``` for

```
epel_repo: https://dl.fedoraproject.org/pub/epel/epel-release-latest-{{ansible_distribution_major_version|int}}.noarch.rpm
```

It be even better to just use the ```yum_repository``` module at this point e.g.

```
yum_repository:
    name: epel
    description: EPEL YUM repo
    baseurl: https://download.fedoraproject.org/pub/epel/epel-release-latest-{{ansible_distribution_major_version|int}}/x86_64/
```
I don't have a pair of CentOS6/RHEL6 VM's handy to test this at the moment so might need some modification/testing.

knechtionscoding commented:
@sadsfae I'm removing the logstash beats port. That was a variable I was using for another use that mistakenly got transferred in. It should be the logstash syslog port. All beats should be using the same port. And you've already got the adaptive ;) firewall code for it.

Sorry, I thought I gave you the fully tested version of this code. Apparently that wasn't the case. Should be up to snuff now.

knechtionscoding commented:
I'm not sure of the need for supporting the installation of ELK on CentOS 6. Sorry I wasn't clearer on the initial PR. I just tested the installation of beats on CentOS 6. I can add CentOS6 support for the installation of ELK but I agree with your initial point regarding it that it is 7 years old and not worth the effort.

sadsfae commented:
Yeah I think staying on RHEL/CentOS7+ is the way to go.

On Fri, Sep 8, 2017 at 2:16 PM, Sven Hans Knecht <notifications@github.com>
wrote:

> I'm not sure of the need for supporting the installation of ELK on CentOS
> 6. Sorry I wasn't clearer on the initial PR. I just tested the installation
> of beats on CentOS 6. I can add CentOS6 support for the installation of ELK
> but I agree with your initial point regarding it that it is 7 years old and
> not worth the effort.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sadsfae/ansible-elk/pull/41#issuecomment-328100363>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADwicqCjFziNtI8X540ebq4W6cTXUEJZks5sgT4jgaJpZM4POqjH>
> .
>



-- 
Will Foster
Engineering R&D ⛅ Red Hat, Inc.
⚓ wfoster ✎ hobo.house ⚿ 07BC7E81

knechtionscoding commented:
Agreed.

sadsfae commented:
@KnechtionsCoding 

Tested this first opting not to install any beat components, just needs a few changes:

- filebeat SSL data wants to be kept in ```/etc/beat/``` instead of ```/etc/elk``` so either change it in the filebeat.yml.j2 or in the filebeat playbook.
- filebeat role for ```elk-client.yml``` is mandatory (as the client needs at least one transport and that's what we support).  I would just remove the ```install_filebeat``` option from ```group_vars/all.yml``` in this case and any conditionals you do off it.  The other optional beat stuff is good though - testing that now.

Looking good!


```
diff --git a/install/elk-client.yml b/install/elk-client.yml
index e769371..1f81a94 100644
--- a/install/elk-client.yml
+++ b/install/elk-client.yml
@@ -8,7 +8,7 @@
   vars_files:
     - group_vars/all.yml
   roles:
-    - { role: filebeat, when: (install_filebeat)}
+    - { role: filebeat }
     - { role: metricbeat, when: (install_metricbeat)}
     - { role: packetbeat, when: (install_packetbeat)}
     - { role: heartbeat, when: (install_heartbeat)}
diff --git a/install/roles/elk_client/tasks/main.yml b/install/roles/elk_client/tasks/main.yml
index eced34c..28c6332 100644
--- a/install/roles/elk_client/tasks/main.yml
+++ b/install/roles/elk_client/tasks/main.yml
@@ -38,4 +38,4 @@
     url=http://{{ elk_server }}:{{ elk_server_ssl_cert_port }}/beat-forwarder.key
     dest=/etc/elk/beat-forwarder.key
   become: true
-  when: (elk_client_ssl_cert_exists != 0)
\ No newline at end of file
+  when: (elk_client_ssl_cert_exists != 0)
diff --git a/install/roles/filebeat/templates/filebeat.yml.j2 b/install/roles/filebeat/templates/filebeat.yml.j2
index aa56b78..e166866 100644
--- a/install/roles/filebeat/templates/filebeat.yml.j2
+++ b/install/roles/filebeat/templates/filebeat.yml.j2
@@ -217,13 +217,13 @@ output:
     # tls configuration. By default is off.
     ssl:
       # List of root certificates for HTTPS server verifications
-      certificate_authorities: [/etc/beat/beat-forwarder.crt]
+      certificate_authorities: [/etc/elk/beat-forwarder.crt]
 
       # Certificate for TLS client authentication
-      certificate: "/etc/beat/beat-forwarder.crt"
+      certificate: "/etc/elk/beat-forwarder.crt"
 
       # Client Certificate Key
-      key: "/etc/beat/beat-forwarder.key"
+      key: "/etc/elk/beat-forwarder.key"
 
       # Controls whether the client verifies server certificates and host name.
       # If insecure is set to true, all server host names and certificates will be
```

knechtionscoding commented:
Okay. I'll make that change now.

And I'll remove the conditional for filebeat then. I was trying to maintain consistency across all the beats, but if that is considered mandatory then I'll remove it.

knechtionscoding commented:
@sadsfae I updated the conditional.

However I'm a little confused on your SSL data comment. Right now I'm storing the ssl cert in /etc/beat. This is because in the eventuality that heartbeat, metricbeat, etc are installed they need to use the same cert. Thus it was renamed. and moved to /etc/beat/. I can't find anywhere where it is pointing to /etc/elk.

On this file: https://github.com/KnechtionsCoding/ansible-elk/blob/refactor-beat-support/install/roles/filebeat/templates/filebeat.yml.j2 it's pointint to /etc/beat (the correct location). Is there somewhere where it isn't?

sadsfae commented:
> However I'm a little confused on your SSL data comment. Right now I'm storing the ssl cert in /etc/beat. This is because in the eventuality that heartbeat, metricbeat, etc are installed they need to use the same cert. Thus it was renamed. and moved to /etc/beat/. I can't find anywhere where it is pointing to /etc/elk.

> On this file: https://github.com/KnechtionsCoding/ansible-elk/blob/refactor-beat-support/install/roles/filebeat/templates/filebeat.yml.j2 it's pointint to /etc/beat (the correct location). Is there somewhere where it isn't?

Your PR has it going to /etc/elk/* for the SSL certificate and key here, this is where I think it needs to be changed to ```/etc/beat/```

file: ```install/roles/elk_client/tasks/main.yml```

```
18  - name: Import ElasticSearch GPG Key
19    rpm_key: key=http://packages.elastic.co/GPG-KEY-elasticsearch
20      state=present
29 - name: Install ELK server SSL client certificate
30   get_url:
31     url=http://{{ elk_server }}:{{ elk_server_ssl_cert_port }}/beat-forwarder.crt
32     dest=/etc/elk/beat-forwarder.crt
33   become: true
34   when: (elk_client_ssl_cert_exists != 0)
35 
36 - name: Install ELK server SSL client certificate
37   get_url:
38     url=http://{{ elk_server }}:{{ elk_server_ssl_cert_port }}/beat-forwarder.key
39     dest=/etc/elk/beat-forwarder.key
40   become: true
41   when: (elk_client_ssl_cert_exists != 0)
```

knechtionscoding commented:
Sorry, too long of a day. That is fixed so it's now pointing to /etc/beat.

sadsfae commented:
@KnechtionsCoding great work so far, looks good on my end. 

Does heartbeat require a listening port (and subsequent iptables/firwalld rule) on the ELK server or is it utilizing the logstash TCP/5044 port ala

https://www.elastic.co/guide/en/beats/heartbeat/current/config-heartbeat-logstash.html

If yes and it doesn't need it's own listening port spun up (and subsequent version of the iptables/firewald stuff) I think it's good to be merged.

knechtionscoding commented:
@sadsfae It does not. It uses the same port as the other beats so that A) there is consistency and B) there is the ability to install it on a different server then the ELK server. 

Heartbeat is unique in that it is only installed in one place on the network and then reaches out. I added this to my elk server, however there is definitely the argument for having it separate if you are in a bigger network. As such it is uniform with the other beats.

sadsfae commented:
@KnechtionsCoding merged now, thank you for your patience and tenacity - it certainly wasn't a trivial enhancement and I really appreciate you submitting it and working through integration.
